Sources/_SwiftSyntaxTestSupport/Syntax+Assertions.swift

```diff
@@ -32,3 +33,192 @@ public func XCTAssertNext<Iterator: IteratorProtocol>(
+  public func assertSameStructure(afterMarker: String? = nil, _ expected: Syntax, includeTrivia: Bool = false,
+                                  file: StaticString = #filePath, line: UInt = #line) throws {
+    let diff = try findFirstDifference(afterMarker: afterMarker, baseline: expected, includeTrivia: includeTrivia)
+    XCTAssertNil(diff, diff!.debugDescription, file: file, line: line)
+  }
+}
+
+public enum SubtreeError: Error, CustomStringConvertible {
+  case invalidMarker(name: String)
+  case invalidSubtree(tree: Syntax, afterUTF8Offset: Int, type: String)
+
+  public var description: String {
+    switch self {
+    case let .invalidMarker(name):
+      return "Could not find marker with name '\(name)'"
+    case let .invalidSubtree(tree, afterUTF8Offset, type):
+      return "Could not find subtree after UTF8 offset \(afterUTF8Offset) with type \(type) in:\n\(tree.debugDescription)"
+    }
+  }
+}
+
+/// Finds all marked ranges in the given text, see `Marker`.
+fileprivate func findMarkedRanges(text: String) -> [Marker] {
+  var markers = [Marker]()
+  while let marker = nextMarkedRange(text: text, from: markers.last?.range.upperBound ?? text.startIndex) {
+    markers.append(marker)
+  }
+  return markers
+}
+
+fileprivate func nextMarkedRange(text: String, from: String.Index) -> Marker? {
+  guard let start = text.range(of: "#^", range: from ..< text.endIndex),
+        let end = text.range(of: "^#", range: start.upperBound ..< text.endIndex) else {
+    return nil
+  }
+
+  let markerRange = start.lowerBound ..< end.upperBound
+  let name = text[start.upperBound ..< end.lowerBound]
+
+  // Expand to the whole line if the line only contains the marker
+  let lineRange = text.lineRange(for: start)
+  if text[lineRange].trimmingCharacters(in: .whitespacesAndNewlines) == text[markerRange] {
+    return Marker(name: name, range: lineRange)
+  }
+  return Marker(name: name, range: markerRange)
+}
+
+fileprivate struct Marker {
+  /// The name of the marker without the `#^` and `^#` markup.
+  let name: Substring
+  /// The range of the marker.
+  ///
+  /// If the marker contains all the the non-whitepace characters on the line,
+  /// this is the range of the entire line. Otherwise it's the range of the
+  /// marker itself, including the `#^` and `^#` markup.
+  let range: Range<String.Index>
+}
+
+fileprivate class SyntaxTypeFinder: SyntaxAnyVisitor {
+  private let offset: Int
+  private let type: SyntaxProtocol.Type
+  private var found: Syntax?
+
+  private init(offset: Int, type: SyntaxProtocol.Type) {
+    self.offset = offset
+    self.type = type
+    self.found = nil
+    super.init(viewMode: ._all)
+  }
+
+  fileprivate override func visitAny(_ node: Syntax) -> SyntaxVisitorContinueKind {
+    if found != nil {
+      return .skipChildren
+    }
+
+    if node.positionAfterSkippingLeadingTrivia.utf8Offset >= offset && node.syntaxNodeType == type {
+      found = node
+      return .skipChildren
+    }
+
```

Comment:
A little performance optimization suggestion (not sure if it makes much of a difference in practice): We don’t need to walk the children if the node’s end location is before `offset`

```swift
if node.endPosition.utf8Offset < offset {
  return .skipChildren
}
```

Reply:
Ah yeah, good point. Done.